### PR TITLE
Enabling op reentrancy enforcement at the DDS/runtime level

### PR DESCRIPTION
**Pull request by andre4i**
See **ADO:2190** and **ADO:1834** for details.

We should not be allowing ops to be submitted while processing another op. This can result in inconsistencies, as the ops may end up having a different reference sequence number which is symptom that the op was produced without having a consistent view of the current state. This may lead to either different clients eventually reaching inconsistent states or to ops being applied FIFO on a single client, changing the expected order of events.

This change will allow (opt-in):
- block submitting ops in the event handler via a new runtime API which throws if a new op is being submitted while the event handling is still executing
- asserts that all ops within a JS turn batch will have the same reference sequence number. If that's not the case, the container will close and the batch will be discarded, in order to minimize the spread to the other clients
- The `AgentScheduler`, which was using this pattern, will now push the task reassignment (which was done within the event handler) to the next batch by forcing a JS turn

To enable, use the `IContainerRuntimeOptions.enableOpReentryCheck` property. The new behavior of SharedObject and the runtime can be disabled at runtime via a feature gate named `Fluid.ContainerRuntime.DisableOpReentryCheck`

### CODE REVIEW

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 1]
```diff
@@ -2777,6 +2788,16 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
...
     }
 
+    private verifyCanSubmitOps() {
+        if (this.mc.config.getBoolean("Fluid.ContainerRuntime.DisableOpReentryCheck") === true) {
+            return;
+        }
+
+        if (this.runWithoutOpsCalls > 0) {
+            throw new UsageError("Op was submitted from within a `runWithoutOps` callback");
+        }
```
Comment (NicholasCouri) [inline packages/runtime/container-runtime/src/containerRuntime.ts:2798]: We will probably break existing scenarios, won't we ? I wonder if we should simply log an event to see how often it happens.

**Comment by taylorsw04** [comment]
Will this prohibit DDS code itself from submitting ops during processing? We actually have several instances of this in SharedTree (and we are careful to avoid any inconsistencies).

File: packages/runtime/container-runtime/src/batchManager.ts [hunk 2]
```diff
@@ -84,4 +89,21 @@ export class BatchManager {
...
     }
+
+    private checkReferenceSequenceNumber(message: BatchMessage) {
+        if (this.mc.config.getBoolean("Fluid.ContainerRuntime.DisableOpReentryCheck") === true) {
+            return;
+        }
+
+        if (this.pendingBatch.length > 0
+            && message.referenceSequenceNumber !== this.pendingBatch[0].referenceSequenceNumber) {
+            throw new GenericError(
```
Comment (taylorsw04) [inline packages/runtime/container-runtime/src/batchManager.ts:100]: Will this prohibit DDS code itself from submitting ops during processing? We actually have several instances of this in SharedTree (and we are careful to avoid any inconsistencies).

**Review by taylorsw04 — CHANGES_REQUESTED** [review]

File: packages/runtime/container-runtime/src/batchManager.ts (hunk 2 above)
Comment (taylorsw04) [reply packages/runtime/container-runtime/src/batchManager.ts:100]: I also leverage this in a DDS from an FHL project that I'll be spinning up a PR for soon, so I'd love to chat about this.

Comment (andre4i) [reply packages/runtime/container-runtime/src/batchManager.ts:100]: Hi @taylorsw04 , can you point me to an example of this in SharedTree? Are the scenarios covered by test and are they in the same monorepo as the client code in this PR? If tests pass then I don't believe this should block your scenarios.

The gist of it is that this validation ensures that all ops in the batch have been created while having the same view of the world (the reference sequence number). On a case-by-case basis, it might be safe or it might not be, hence it should be regarded as undefined behavior and the proposal here is to block it when the op is produced.

> we are careful to avoid any inconsistencies

I'm interested in how this is accomplished. Can you point me to the relevant logic?

FYI @vladsud

Comment (taylorsw04) [reply packages/runtime/container-runtime/src/batchManager.ts:100]: I believe they happen to pass because they spin up async work (blob uploading) before sending that off in an op. This is true in all cases that I know of, FWIW, but I don't think that's worth focusing on as I think there are viable reasons that a DDS can shoulder the burden of handling this case (although you might have ideas on how to accomplish the same, but safer).

An example: https://github.com/microsoft/FluidFramework/blob/main/experimental/dds/tree/src/SharedTree.ts#L1280

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 3]
```diff
@@ -523,7 +523,18 @@ export abstract class SharedObjectCore<TEvent extends ISharedObjectEvents = ISha
      * @returns `true` if the event had listeners, `false` otherwise.
      */
     public emit(event: EventEmitterEventType, ...args: any[]): boolean {
-        return this.callbacksHelper.measure(() => super.emit(event, ...args));
+        return this.callbacksHelper.measure(() => {
+            let result = false;
+
+            // Creating ops while handling a DDS event can lead
+            // to undefined behavior. The runtime must enforce op
```
Comment (vladsud) [inline packages/dds/shared-object-base/src/sharedObject.ts:530]: I'd add more concrete data here, like the fact that most DDSs (map included) would send ops in reverse order, in case of reentrancy while generating nested local edits.

File: packages/runtime/container-runtime/src/batchManager.ts [hunk 4]
```diff
@@ -84,4 +89,21 @@ export class BatchManager {
             },
         };
     }
+
+    private checkReferenceSequenceNumber(message: BatchMessage) {
+        if (this.mc.config.getBoolean("Fluid.ContainerRuntime.DisableOpReentryCheck") === true) {
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/batchManager.ts:94]: Can host set feature gate in their code to ensure it's disabled, if they want to? I'm thinking of scenario where a customer (including 3rd party customer) picks up build with these changes and learns they can't move forward. I.e. it's possible we will learn there are legit cases, and we have to back off with our enforcement, but it would be nice not to have to issue patches to do so.

File: packages/runtime/container-runtime/src/batchManager.ts (hunk 2 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/batchManager.ts:100]: @taylorsw04, just to be sure - are you saying it's Ok to push that requirement to DDSs not to issue ops while processing ops? Or reverse?

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 5]
```diff
@@ -826,6 +826,22 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
...
+     * Invokes the given callback and expects that no ops are submitted
+     * until execution finishes. If an op is submitted, an error will be raised.
+     *
+     * Can be disabled by feature gate `Fluid.ContainerRuntime.DisableOpReentryCheck`
+     *
+     * @param callback - the callback to be invoked
+     */
+    public runWithoutOps(callback: () => void): void {
+        this.runWithoutOpsCalls++;
+        callback();
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:841]: It does not handle exceptions well - this.runWithoutOpsCalls-- should be in finally clause of try/catch

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 6]
```diff
@@ -2744,6 +2751,16 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
         }
     }
 
+    private verifyCanSubmitOps() {
+        if (this.mc.config.getBoolean("Fluid.ContainerRuntime.DisableOpReentryCheck") === true) {
+            return;
+        }
+
+        if (this.runWithoutOpsCalls > 0) {
+            throw new UsageError("Op was submitted from within a `runWithoutOps` callback");
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:2760]: That error alone is not detailed enough to help most customers understand what happened and what they need to do.
I think we need to find a way (either via comments in the code, readme.md in the package, something else, or combo of all the above) to let OCEs and developers (internal and external) know what to do next.

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 7]
```diff
@@ -1001,38 +1018,38 @@ export const mixinRequestHandler = (
...
-    handler: (runtime: FluidDataStoreRuntime) => Promise<{ path: string[]; content: string; } | undefined >,
+    handler: (runtime: FluidDataStoreRuntime) => Promise<{ path: string[]; content: string; } | undefined>,
     Base: typeof FluidDataStoreRuntime = FluidDataStoreRuntime,
 ) => class RuntimeWithSummarizerHandler extends Base {
-        private addBlob(summary: ISummaryTreeWithStats, path: string[], content: string) {
-            const firstName = path.shift();
-            if (firstName === undefined) {
-                throw new LoggingError("Path can't be empty");
-            }
+    private addBlob(summary: ISummaryTreeWithStats, path: string[], content: string) {
```
Comment (vladsud) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:1024]: I'd suggest to submit formatting changes in separate PRs. It's makes tracking history harder

File: packages/runtime/container-runtime/src/batchManager.ts (hunk 2 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/batchManager.ts:100]: If re-entrancy is allowed, then even if DDS is very careful here and does "all the right things", an event callback causing re-entrancy results in callbacks running in reverse order for _some_ users of the system. This is very dangerous for correctness.

I.e. image situation that there are 2 callbacks registered - A & B.
If on change # 1, callback A runs first and makes another change # 2, then we will see this flow:

A for # 1
A for # 2
B for # 2
B for # 1

B is getting events in wrong order.

**Comment by vladsud** [comment]
Overall looks good to me (even though I'm not a big fan of propagating new API across all layers).
But I'd suggest to completely ship it dark, and then ask various partners to enable it in their repos and provide feedback.
And then, once we have good feeling about where it's headed, change the gate (and its name) to be kill-bit switch in next release.
To some extent, I think it should be in container runtime config, I think, pushing requirements to flight (if clients want so) to clients. Feature gates are mostly used to avoid that burden on all customers, but I think this case is the one where we should not be afraid of this extra burden and be more direct in exposing it and asking for feedback

File: packages/runtime/container-runtime/src/batchManager.ts (hunk 4 above)
Comment (andre4i) [reply packages/runtime/container-runtime/src/batchManager.ts:94]: I'll take your advice and make this opt-in via the runtime options with the feature gate as a break-glass solution.

File: packages/runtime/datastore/src/dataStoreRuntime.ts (hunk 7 above)
Comment (andre4i) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:1024]: I agree with you, but it gets annoying to enable/disable code formatting in vscode for files which are out of sync. Perhaps we should format everything and make this a linter error.

File: packages/runtime/container-runtime/src/batchManager.ts (hunk 2 above)
Comment (taylorsw04) [reply packages/runtime/container-runtime/src/batchManager.ts:100]: @vladsud I am saying the reverse. I do think your scenario is dangerous, and should probably never be implemented. However, in the case of our DDSs, we are submitting ops that are only mutating internal state and never invoke callbacks (both are related to persisting state via uploading blobs). So, I agree that event-calling re-entrancy is likely _always_ bad, but we aren't in that bucket.

Comment (andre4i) [reply packages/runtime/container-runtime/src/batchManager.ts:100]: The behavior is now opt-in, so by default this is allowed. So at least for now, the default will allow sending ops from event handlers. Even if explicitly disallowed, there's a feature flag which can override it at runtime. So, in any case, your scenario will not be blocked, regardless of how we go forward (as you'd always have an option to turn it off, either from code or from feature flags - preferably the former).

However, I'm interested in your scenario and how would it differ from the example Vlad described. When you say internal state, do you mean the state of another DDS or the current DDS? Or state which doesn't fire another event handler?

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 1 above)
Comment (andre4i) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2802]: This is opt-in now, so the point is moot.

File: packages/runtime/container-runtime/src/batchManager.ts (hunk 2 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/batchManager.ts:100]: We could make change design and work off events only. I.e. allow any op sending, but disallow re-entrancy in events across container runtime. But I think that will cut off scenario that might be useful - having an "uber-DDS" (or "channel") that is an aggregation of multiple child DDSs. In such system, lower level changes (and events) are translated to higher level events on the parent, and it's Ok to have event reentrancy.

So, it feels like catching these issues at the op submission level is a bit better approach. And so far, it sounds like we have not seen an example where pushing people to do async op submission is a problem.

File: api-report/datastore.api.md [hunk 8]
```diff
@@ -70,6 +70,7 @@ export class FluidDataStoreRuntime extends TypedEventEmitter<IFluidDataStoreRunt
     dispose(): void;
     // (undocumented)
     get disposed(): boolean;
+    ensureNoDataModelChanges(callback: () => void): void;
```
Comment (vladsud) [inline api-report/datastore.api.md:73]: This is probably out of scope for this change, but consider defining "base" interface that container runtime, data store context and data store runtime needs to implement, as a way to start moving toward unified channels :)

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 9]
```diff
@@ -826,6 +839,25 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
...
+
+    /**
+     * Invokes the given callback and expects that no ops are submitted
+     * until execution finishes. If an op is submitted, an error will be raised.
+     *
+     * Can be disabled by feature gate `Fluid.ContainerRuntime.DisableOpReentryCheck`
+     *
+     * @param callback - the callback to be invoked
+     */
+    public ensureNoDataModelChanges(callback: () => void): void {
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:852]: It is more useful to have ensureNoDataModelChanges<T>(callback: () => T): T;

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 10]
```diff
@@ -2752,6 +2780,28 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
         }
     }
 
+    private verifyCanSubmitOps() {
+        if (!this.enableOpReentryCheck) {
+            return;
+        }
+
+        if (this.ensureNoDataModelChangesCalls > 0) {
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:2788]: I think it's better to have

if (this.enableOpReentryCheck && this.ensureNoDataModelChangesCalls > 0) throw(...)

I.e. in one IF statement

**Review by vladsud — APPROVED** [review]

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 9 above)
Comment (andre4i) [reply packages/runtime/container-runtime/src/containerRuntime.ts:852]: Ah.. Great point!

File: api-report/datastore.api.md (hunk 8 above)
Comment (andre4i) [reply api-report/datastore.api.md:73]: noted!

File: packages/runtime/container-runtime/src/batchManager.ts (hunk 2 above)
Comment (taylorsw04) [reply packages/runtime/container-runtime/src/batchManager.ts:97]: @andre4i a good example of this was the flushing to storage that the partial map DDS does in response to receiving an op (on some cadence). The key here is that it will never invoke event handlers and is only persisting a stable view of the internals of the DDS (no reliance on other DDS' states).

Regardless, I'm happy with the changes.

**Review by taylorsw04 — APPROVED** [review]
